File: src/main/java/com/example/coupon/coupon/domain/entity/Coupon.java
package com.example.coupon.coupon.domain.entity;

import jakarta.persistence.*;
import lombok.*;

import java.time.LocalDateTime;

@Builder
@AllArgsConstructor
@NoArgsConstructor(access = AccessLevel.PROTECTED)
@Getter
@Table(name = "coupon")
@Entity
public class Coupon {

    @Column(name = "coupon_id")
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    @Id
    private Long couponId;

    @Column(name = "name")
    private String name;

    @Column(name = "quantity")
    private int quantity;

    @Column(name = "created_at")
    private LocalDateTime createdAt;
}

File: src/main/kotlin/com/example/couponservice/domain/Coupon.kt
package com.example.couponservice.domain

import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id

@Entity
class Coupon(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,

    @Column(nullable = false, unique = true)
    val name: String,

    @Column(nullable = false)
    var issuedCount: Int = 500
) {
    fun issue() {
        if (issuedCount <= 0) {
            throw RuntimeException("Coupon is not available. couponId: $id")
        }

        issuedCount -= 1
    }
}